### DISCUSSION TITLE
Prefetch on hover/on touch option?

### DISCUSSION THREAD

**Discussion by vvo** [vercel/next.js#11793] [Help] [answered]
Hi there, I was "surprised" to learn that next/link [will prefetch any Link that is in the viewport](https://nextjs.org/docs/api-reference/next/link). For two reasons:
1- When you have a long menu list, it seems counter-intuitive to load it all at once
2- I thought that prefetching could instead be based on whether or not the user is likely willing to click on that link

I just want to open the discussion and see if you ever thought/discussed on implementing a prefetch that is based on mouse over/touch instead? Just like https://instant.page/ is doing.

cc @dieulot maybe since you're the author of instant.page you might have some ideas here!

**Comment by dieulot** [comment]
Assuming they don’t get a lot of complaints from their viewport prefetching, I recommend them to switch to prefetching on hover on desktop and keep viewport prefetching on mobile. It would make a lot less requests and be nearly as fast as using the viewport on all form factors.

If Next got complaints about being a server hog, also switch to touchstart prefetching on mobile.

**Comment by abdou19-97** [comment]
Hi everyone!
I need help with creating a BMI web page for the my assignment!!

**Reply by vvo** [reply]
Hey @abdou19-97 you will get more help from https://stackoverflow.com/questions/tagged/next.js?tab=Newest

**Comment by locomoco28** [comment]
I had realized this when I setup my first project in NextJS a few days ago. I exported my project to my website as static files and realized in the network tab that it's pre-fetching my pages when I hover my navigation links and fixed it by creating my own Link component like this
```javascript
import Link from 'next/link';

export default (props) => (
    <Link
        href={props.href}
        as={props.as}
        passHref={props.href}
        prefetch={props.prefetch || false}
        replace={props.replace}
        scroll={props.scroll}
    >
        <a {...props}>{props.children}</a>
    </Link>
);
```
This also makes it easier to add any attributes you want to add to the `<a/>` element :)

**Comment by balazsorban44** [comment]
@vvo I assume you haven't found a solution yet?

I have a similar problem, where I show around 20-30 links at the same time in the viewport, each of them pointing to a page themselves. It fires 20-30 requests immediately, and every network traffic freezes.

I would like to create a Link whose prefetch behavior could be controlled by hover.

What I have so far, is the following hook:

```ts
import * as React from "react"
/**
 * @note Treat `true` as `undefined`. See explanation:
 * @see https://github.com/vercel/next.js/blob/master/errors/prefetch-true-deprecated.md
 */
export default function usePrefetchOnHover(prefetch?: boolean | "onHover") {
  const [shouldPrefetch, setShouldPrefetch] = React.useState<false | undefined>(false)
  if (prefetch === "onHover") {
    return {
      prefetch: shouldPrefetch,
      onMouseOver: () => setShouldPrefetch(undefined),
      onFocus: () => setShouldPrefetch(undefined),
    }
  }
  return { prefetch: prefetch ? undefined : prefetch }
}
```

This, if given `onHover` as a parameter would make the prefetch action controllable. The result of the hook could be spread onto a `next/link` component. But it does not seem to work.


```jsx
const Link = ({prefetch, children, ...props}) => {
  return (
    <NextLink {...usePrefetchOnHover(prefetch)}/>
      <a>{children}</a>
    </NextLink>
  )
}


<NextLink prefetch="onHover" href="/some/page">
  Some page
</NextLink>
```

**Comment by TommySorensen** [comment]
This is working out of the box from NextJS [v10.0.3](https://github.com/vercel/next.js/releases/tag/v10.0.3).
When you set `prefetch={false}` Next will prefetch the url on `onMouseEnter`.
```js
<Link prefetch={false}>
```

PR that fixed the issue: https://github.com/vercel/next.js/pull/18859

**Reply by vvo** [reply]
Thanks @TommySorensen, should we update the documentation here: https://nextjs.org/docs/api-reference/next/link? Or is there a place where it's said that prefetch=false will prefetch on mouse enter?

**Reply by TommySorensen** [reply]
The documentation does not describe this optimization at all, so yeah. Someone should do at PR with some text about this 👍
I guess this is one of those hidden NextJS features we should not think about. Maybe that's why it's not described 🙂

**Reply by steven-tey** [reply]
Hell yeah just tried it out – checked my server logs and this checks out! We need to get this PR'ed into the docs ASAP – it's a wonderful fix that creates an incredible user experience! Reminds me of [Instant Page](https://instant.page/).

**Reply by vvo** [reply]
Update: this is now inside the docs, closing.

**Comment by sakkelaaksonen-siili** [comment]
prefetch={false} should not prefetch anything on any event, be it mounting or hovering.

I suggest something like this: prefetch=always | hover | none or false or undefined.
I can crash a server by simply moving a mouse over a menu....

**Reply by persil** [reply]
True. This is sick. It is even re-fetching json on every mouseOver event - you can move your mouse back and forth 1 px and just see how the transferred resources rises in the devTools - in my case this is also not acceptable as I try to use SSR to minimize server load over the long run, not the opposite.

**Reply by trompx** [reply]
For ecommerce mega-menus, sidebar menus, this is totally unusable.

**Comment by vvo** [comment, marked as answer]
Update: you can prefetch on hover with prefetch={false} (good or bad).

See here: https://nextjs.org/docs/api-reference/next/link

**Reply by eliawk** [reply]
and how to disable the prefetch on hover too?

**Reply by NawarA** [reply]
@vvo see my comment https://github.com/vercel/next.js/discussions/11793#discussioncomment-3928305

**Comment by airidasj** [comment]
I too need to completely disable prefetching, as it clogs the server for larger e-commerce sites.

**Comment by frukc** [comment]
To force prefetching on mouseOver even if the flag is set to false is terrible practice. Only way to rationalize this is looking at the business model of **_Vercel_**, they provide hosting services and charge based on data transferred...

**Reply by sakkelaaksonen-siili** [reply]
Indeed. Too many of these and people will just **fork it on github** , fix the bs features , call it and-next-after-thatJs and effectively undermine the business model.

**Comment by arm-channarong** [comment]
but on lighthouse prefetch will still request when i set prefetch to false and it's reduce the performance score how can i disable prefetch?

**Comment by wenche** [comment]
We see a lot of 404 with this feature. 

Our use case is a language selector, where you can switch to the corresponding page in another language. In our case `EN | NO`. When I'm on the Norwegian version of like `/no/om-oss` and hover the link to 'EN', Next.js will "help" me and prefetch _both_ `/en/about-us.json` and `/no/about-us.json`. The last one does of course not exist since our URLs are translated and the result is a 404 which is bloating our logs. :/

**Comment by dmitrypasyutin** [comment]
IMO, prefetch={false} should disable prefetching at all. Users may hang up your server by simply hovering the mouse all over the page that contains a lot of links. Please, at least consider caching the prefetch response, but the best solution would be avoid prefetching when not needed.

**Comment by NawarA** [comment]
I'd like to also vote for `prefetch={false|hover|true}`

where `true` is the default behavior we have today; `hover` is prefetch on link hover (what today is called false), and `false` is completely  disabled.

this would give devs full control

alternative naming options: `prefetch={"all"|"hover"|"off"}`, where default behavior is "all"

**Reply by Netail** [reply]
Would rename "all" to "viewport" perhaps, as the default behaviour is based on if the link enters the viewport

**Reply by Netail** [reply]
As this discussion has been closed and not in the correct category, we can create a new "ideas" discussion with this proposal

EDIT: https://github.com/vercel/next.js/discussions/77877 seems to cover this already

**Comment by vsnthdev** [comment]
I also vote on @NawarA's way of implementation.

Whether to `prefetch` or not should be decided solely by the developer not the framework.

**Comment by 2-fly-4-ai** [comment]
Can someone please implement this? It's absolute madness to apply prefetch on hover as the default when prefetch is set to false. Just think about it for a second, when you scroll down a page the cursor passes over a lot of links. I used up my entire budget of serverless functions within ten days because of this default behavior.

**Reply by trompx** [reply]
This has been answered in [another post](https://github.com/vercel/next.js/discussions/24437#discussioncomment-3752350), but this is still not enough customizable. In the case of prefetch hover, it would also be great to be able to **control the delay after which the prefetch is triggered on hover**, as I mentioned in my answer to timneutkens.

**Reply by lucasbasquerotto** [reply]
I implemented a custom link that never prefetches on hover. You can see the implementation and how to use in the following comment: https://github.com/vercel/next.js/discussions/11793#discussioncomment-4710408

**Reply by trompx** [reply]
I got a bunch of email notifications from your answers @2-fly-4-ai, but can't see them appears here, have you removed them yourself? I also feel your pain and we shouldn't have to resort to some "hacks" to control the prefetch behavior.

**Comment by lucasbasquerotto** [comment]
I wanted my links to never prefetch, even on hover. So, I created a custom link to use for my needs. I use it like:

```tsx
const MyLink = () => (
    <MainLink href="/path/to/page">
        Click Me
    </MainLink>
)
```

Below is a reference of my implementation (with a few changes):

```tsx
import { useRouter } from 'next/router';
import type { HTMLAttributeAnchorTarget } from 'react';
import React from 'react';

export interface MainLinkProps {
    href?: string;
    target?: HTMLAttributeAnchorTarget | undefined;
    className?: string;
    children: React.ReactNode;
}

const MainLink: React.FC<MainLinkProps> = React.forwardRef(
    (
        { href, target, className, children }: MainLinkProps,
        ref: React.Ref<HTMLAnchorElement> | undefined,
    ) => {
        const router = useRouter();

        const handleClick = React.useCallback(
            async (e: React.SyntheticEvent) => {
                if (href) {
                    e.preventDefault();
                    return await router.push(href);
                }
            },
            [router, href],
        );

        return (
            <a
                ref={ref}
                className={className}
                target={target}
                href={href || '/'}
                onClick={handleClick}
            >
                {children}
            </a>
        );
    },
);

export default MainLink;
```

**Reply by 2-fly-4-ai** [reply]
> I wanted my links to never prefetch, even on hover. So, I created a custom link to use for my needs. I use it like:
> 
> ```tsx
> const MyLink = () => (
>     <MainLink href="/path/to/page">
>         Click Me
>     </MainLink>
> )
> ```
> 
> Below is a reference of my implementation (with a few changes):
> 
> ```tsx
> import { useRouter } from 'next/router';
> import type { HTMLAttributeAnchorTarget } from 'react';
> import React from 'react';
> 
> export interface MainLinkProps {
>     href?: string;
>     target?: HTMLAttributeAnchorTarget | undefined;
>     className?: string;
>     children: React.ReactNode;
> }
> 
> const MainLink: React.FC<MainLinkProps> = React.forwardRef(
>     (
>         { href, target, className, children }: MainLinkProps,
>         ref: React.Ref<HTMLAnchorElement> | undefined,
>     ) => {
>         const router = useRouter();
> 
>         const handleClick = React.useCallback(
>             async (e: React.SyntheticEvent) => {
>                 if (href) {
>                     e.preventDefault();
>                     return await router.push(href);
>                 }
>             },
>             [router, href],
>         );
> 
>         return (
>             <a
>                 ref={ref}
>                 className={className}
>                 target={target}
>                 href={href || '/'}
>                 onClick={handleClick}
>             >
>                 {children}
>             </a>
>         );
>     },
> );
> 
> export default MainLink;
> ```

Would this have any noticeable effect on SEO? Trying to track down why my ranking took such a massive plunge after applying this. It could also be the insane prefetch behavior and errors generated from it that just sent it into an SEO death Spiral though. I have seen some documentation suggestions to use passref with custom link components. Anyways thanks for the code, it saved me from pulling the plug altogether on this tech stack.  

Lost all keywords on site and 1000x fewer impressions. Difficult to track down the exact cause, and these things take time to rebound. So just trying to cover my bases. Let me know what you think if this would make any difference or if there's anything else to consider regarding the Linking I should consider.

**Reply by lucasbasquerotto** [reply]
@2-fly-4-ai I can't say for sure. I don't see why it would cause problems with SEO tough. You could change the component temporarily to use the Next link component under the hood and see if solves the SEO issue. 

As you can see, the anchor component has `href` so I would assume it wouldn't have bad SEO implications (although I can't say for sure when compared with a non-SPA website that loads a new page on navigation, entirely (the HTML) from the server, which is also not the case with Next link, that works similar to my component). 

The Next link component source code seems to be at https://github.com/vercel/next.js/blob/canary/packages/next/src/client/link.tsx if you want to know its internals.

**Comment by psiho** [comment]
I fail to understand this hover prefetching behavior. I cannot understand how can anybody justify this many requests to server just by moving a mouse!? What's the point of optimization when my static page with long lists just shoot hundreds of requests to server? Can somebody please explain it to me because I'm just about to give up and move back to node/express.

**Reply by cvbuelow** [reply]
https://github.com/vercel/next.js/discussions/40268#discussioncomment-3572642

**Comment by 2-fly-4-ai** [comment]
"I believe it is important for the documentation to clearly state that this technology is not suitable for building e-commerce websites or any platform with a high number of links. While there are methods to bypass this issue, it is problematic out of the box. I experienced an unexpected charge due to serverless functions exceeding the limit by a small margin, even after removing my projects from Vercel. This suggests that it is part of their profit strategy to exploit inexperienced developers who may not anticipate this behavior. This experience has been very disheartening and has made me reconsider my investment in this technology. It is not about the cost, but the behavior that makes it unsuitable for these types of applications.

**Comment by khuezy** [comment]
When I set `prefetch={false}`, it doesn't prefetch on hover.

**Reply by jjunior96** [reply]
> When I set `prefetch={false}`, it doesn't prefetch on hover.

It still goes on like this...

**Reply by wimbarelds** [reply]
Same, with `prefetch={false}`, hover does not cause any prefetch.

**Reply by icyJoseph** [reply]
I think, that if you have `legacyBehavior` combined with a custom Anchor tag, like:

```tsx
import Link from "next/link";
import { ComponentPropsWithoutRef, forwardRef } from "react";

// `onClick`, `href`, and `ref` need to be passed to the DOM element
// for proper handling
const MyButton = forwardRef<HTMLAnchorElement, ComponentPropsWithoutRef<"a">>(
  function MyButton({ onClick, href }, ref) {
    return (
      <a href={href} onClick={onClick} ref={ref}>
        Click Me
      </a>
    );
  }
);

function Home() {
  return (
    <Link href="/about" passHref legacyBehavior prefetch={false}>
      <MyButton />
    </Link>
  );
}

export default Home;
```

Then it is not prefetching, unless you spread the props in `MyButton`, or rather, pass the `onMouseEnter` prop. I am still investigating though, but I've seen it misbehave like this in production.

**Comment by adamJLev** [comment]
Vercel loves nextjs "features" [that make them money](https://vercel.com/docs/pricing/networking#edge-requests), like this one that triggers dozens or even hundreds of GET requests on every page load to prefetch links that the user will never click on. 

I think as a developer we would really appreciate a way to override the default so that its `prefetch=false` everywhere, since most websites have tons of links on every page, and users are not going to click _on all_ of them.

**Reply by icyJoseph** [reply]
> Vercel loves nextjs "features" [that make them money](https://vercel.com/docs/pricing/networking#edge-requests), like this one that triggers dozens or even hundreds of GET requests on every page load to prefetch links that the user will never click on. 
> 
> I think as a developer we would really appreciate a way to override the default so that its `prefetch=false` everywhere, since most websites have tons of links on every page, and users are not going to click _on all_ of them.

Some of these things have historical context.

Let's just start with, https://nextjs.org/docs/app/api-reference/components/link#prefetch now shows:

> false: Prefetching will never happen both on entering the viewport and on hover

And some historical context, for hover prefetch, which is a feature, like many others that originated from the web community, several years ago. For example InstantClick, which is used by the engine powering dev.to:

- http://instantclick.io/
- https://github.com/search?q=repo%3Aforem%2Fforem+instantclick&type=code&p=2

I'd agree that, the false option should've been like that, way before... But the hover prefetch, is arguably not that bad. Prefetching on entering the viewport was rather clear in the old docs though. And yes, it can be really annoying to suddenly have so many hits coming to your site.

**Reply by adamJLev** [reply]
thanks for the context @icyJoseph 
I think `prefetch='hover'` would be a welcome addition for most people, maybe even as a default 👌🏼

**Comment by jenswaern** [comment]
I've solved prefetch on hover by just having prefetch as a state and setting it to true on MouseEnter.

```
  const [prefetch, setPrefetch] = useState(false);
  return (
    <Link
      href={href}
      target={target}
      rel={rel}
      prefetch={prefetch}
      onMouseEnter={() => setPrefetch(true)}
    >
      {children}
    </Link>
  );
```
